src/lib.rs

```diff
@@ -53,18 +58,22 @@ where
   ro_consts_circuit_secondary: ROConstantsCircuit<<G1 as Group>::Base>,
   r1cs_gens_secondary: R1CSGens<G2>,
   r1cs_shape_secondary: R1CSShape<G2>,
-  c_primary: C1,
-  c_secondary: C2,
   params_primary: NIFSVerifierCircuitParams,
   params_secondary: NIFSVerifierCircuitParams,
+  c_primary: C1,
+  c_secondary: C2,
+  a1: Option<PoseidonConstants<<G1 as Group>::Scalar, A1>>,
```

Comment:
I think we should be using `ROConstants` instead of `PoseidonConstants` (there was a TODO at the top of the file to completely avoid including poseidon.rs from lib.rs). In the short term, if the goal is to support a few arities of interest (e.g., 2 for VDFs and 6 for Lurk), could we just add those aritites to the hard-coded constants that we generate? This will likely avoid a lot of complexity. What do you think?

Reply:
We could then make implementations of `ROConstants` be able to handle arbitrary arities via the sponge construction.

Reply:
I agree with you on the *general* topology, and I have pushed a commit which I think finally addresses the issues. I moved the arity poseidon constants into the verifier circuit. I didn't put them in `ROConstants` directly for several related reasons:

1. If we retain generality, that would require putting an `Arity` type parameter everywhere, including places it didn't belong.
2. If we discard generality, it would actually have added a lot of complexity and cost around 'dynamic allocation' of specific constants.
3. They're not *really* part of the RO anyway. As implemented here, they're a helper for the application function F.

In any case, my recent commit now successfully hides the PoseidonConstants from the caller almost completely. In the case that a caller wants to implement the `ComputeStep` trait for a non-unary function, they need to handle (but not generate) constants in the `compute_inner` method. This can probably also be eliminated, but I think it's livable for now.